### ISSUE TITLE
I/O starvation when sending transactions with high QPS

### ISSUE THREAD

**Issue by vita-dounai** [hyperledger/caliper#505]
## Context
I used fixed-rate rate controller for my benchmark, but somtimes the Caliper clients seem like falling in sleep and stopping sending transactions. After a long period of sleep, the Caliper clients wake up suddenly and send all blocked transactions at once.

## Expected Behavior

The bandwidth utilization when fixed-rate clients sending transaction should be steady

## Actual Behavior

![image](https://user-images.githubusercontent.com/44284897/61100146-a06c9b80-a497-11e9-883b-33358ea35d58.png)

## Possible Fix

See the code below which is excerpted from 
`runFixedNumber` in `caliper-local-client.js`:
```
 let promises = [];
 while(this.txNum < msg.numb) {
     promises.push(cb.run().then((result) => {
         this.addResult(result);
         return Promise.resolve();
     ));
     await rateControl.applyRateControl(this.startTime, this.txNum, this.results, this.resultStats);
}
```
High QPS would create a lot of Promise tasks, which is prior to I/O tasks. Promise tasks would preempt CPU and I/O tasks have no chance to be executed. For more detail about principle, please refer to this article: [Node.js: How even quick async functions can block the Event-Loop, starve I/O](https://snyk.io/blog/nodejs-how-even-quick-async-functions-can-block-the-event-loop-starve-io/)

## Steps to Reproduce

1. Install Caliper and Caliper CLI
2. Configure to use fixed-rate controller in a round of benchmark 
3. Configure a high tps (> 10000) to the fixed-rate controller.
4. Run benchmark, and monitor network usage.

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-caliper)
- [ ] [GitHub Issues](https://github.com/hyperledger/caliper/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/caliper)

## Context

The sending rate is not fixed! When Clliper clients falling into sleep, my SUT will fall into starvation too, I can't get a accurate result of performance evaluation.
My solution is:
```
        let promises = [];
        while(this.txNum < msg.numb) {
            // If this function calls cb.run() too quickly, micro task queue will be filled with unexecuted promises,
            // and I/O task(s) will get no chance to be execute and fall into starvation, for more detail info please visit:
            // https://snyk.io/blog/nodejs-how-even-quick-async-functions-can-block-the-event-loop-starve-io/
            await this.setImmediatePromise(() => {
                promises.push(cb.run().then((result) => {
                    this.addResult(result);
                    return Promise.resolve();
                }));
            });
            await rateControl.applyRateControl(this.startTime, this.txNum, this.results, this.resultStats);
        }
```

The definition of `setImmediatePromise` is:
```
    /**
     * Put a task to immediate queue of NodeJS event loop
     * @param {function} func The function needed to be exectued immediately
     * @return {Promise} Promise of execution
     */
    setImmediatePromise(func) {
        return new Promise((resolve) => {
            setImmediate(() => {
                func();
                resolve();
            });
        });
    }
```

Although it hurts perfomance of Caliper clients a little, but the sending speed becomes really ***fixed***:
![image](https://user-images.githubusercontent.com/44284897/61100812-0b1ed680-a49a-11e9-9879-93f1815d9357.png)

## Your Environment

* Caliper version: v0.1.0
* NodeJS version: v8.16.0, Docker version: v18.09.7
* Operating System and version: CentOS server, version: 7
* Link to your project: https://github.com/vita-dounai/caliper/tree/fisco-bcos

**Comment by aklenik** [comment]
@vita-dounai Nice analysis and article reference 👍 How many clients did you use to generate the load?
 
(NB: Caliper is still under active development, thus we haven't got to the performance optimization part yet since the architecture and control flow might change in the [near] future.)

**Comment by vita-dounai** [comment]
> @vita-dounai Nice analysis and article reference 👍 How many clients did you use to generate the load?
> 
> (NB: Caliper is still under active development, thus we haven't got to the performance optimization part yet since the architecture and control flow might change in the [near] future.)

8 clients, and the QPS is limited to 10000.

**Comment by aklenik** [comment]
@vita-dounai Hmm, that's 1000+ TPS per client, might be a little too much for a single-threaded Nodejs process. At that point, it even affects the precision of the rate controller. At a 1k TPS, the rate controller must enable a new transaction every ms, but it won't get the execution control that frequently. 

I usually limit the client-wise TPS to around 60-70 TPS, which means using a lot of clients (30+) for high TPS loads, resulting in more precise rate control.

I'm interested to see how your plots change if you use let's say 20-30 clients :)

**Comment by vita-dounai** [comment]
> @vita-dounai Hmm, that's 1000+ TPS per client, might be a little too much for a single-threaded Nodejs process. At that point, it even affects the precision of the rate controller. At a 1k TPS, the rate controller must enable a new transaction every ms, but it won't get the execution control that frequently.
> 
> I usually limit the client-wise TPS to around 60-70 TPS, which means using a lot of clients (30+) for high TPS loads, resulting in more precise rate control.
> 
> I'm interested to see how your plots change if you use let's say 20-30 clients :)

Maybe your solution can work out the problem. But, the reason why I choose to use 8 clients is that the throughput  of my blockchain can reach to 7K~8K TPS, so I need make the test pressure to be 10K QPS to ensure that my blcokchain can be saturated. If a client can only send transactions at QPS < 100 , that means I need launch over 100 clients in my system, which is not reasonable.

Let's consider a more extreme occasion: if a SUT's throughput rate can reach to 100K or more TPS, what should we do? Just add more clients?

BTW, I have tested my blockchain with 32 clients, the phenomenon is still exist, and in addition to this，the overall QPS dropped/(ㄒoㄒ)/~~ Maybe too many active client processes make my operating system running slowly.

**Comment by aklenik** [comment]
@vita-dounai True, your solution is the way to go, I was just interested in the effect on a larger scale. 
Feel free to open another PR with this enhancement, your send rate performance boost is awesome!

**Comment by aklenik** [comment]
Resolved by #509
